core/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -471,19 +472,24 @@ private static void parseObjectOrField(ParseContext context, Mapper mapper) thro
 
     private static ObjectMapper parseObject(final ParseContext context, ObjectMapper mapper, String currentFieldName) throws IOException {
         assert currentFieldName != null;
-        context.path().add(currentFieldName);
 
         ObjectMapper update = null;
         Mapper objectMapper = getMapper(mapper, currentFieldName);
         if (objectMapper != null) {
+            context.path().add(currentFieldName);
             parseObjectOrField(context, objectMapper);
+            context.path().remove();
         } else {
-            ObjectMapper.Dynamic dynamic = dynamicOrDefault(mapper, context);
+
+            // TODO: why Strings.splitStringToArray instead of String.split?
+            final String[] paths = Strings.splitStringToArray(currentFieldName, '.');
+            currentFieldName = paths[paths.length - 1];
+            Tuple<Integer, ObjectMapper> parentMapperTuple = getDynamicParentMapper(context, paths, mapper);
```

Comment:
I tried the above but hit a problem: if the `getDynamicParentMapper` method does not add the dynamically created parents to the context then the nested dynamic objects aren't created correctly. Imagine adding the long field `foo.bar.baz` dynamically. The `getDynamicParentMapper` method will first create the `foo` object field and then the `bar` object field, but because `foo` was not added to the context path when it was added (since this has been moved outside of this method) the `bar` field is added to the root object mapper instead of as a inner object to the `foo` object field.